test/e2e/gardener/shoot/internal/inclusterclient/inclusterclient.go

```diff
@@ -0,0 +1,281 @@
+		cleanupCtx, cancel := context.WithTimeout(context.Background(), time.Minute)
+		defer cancel()
+
+		for _, obj := range objects {
+			Expect(client.IgnoreNotFound(c.Delete(cleanupCtx, obj))).To(Succeed(), "should delete %T %q", obj, client.ObjectKeyFromObject(obj))
+		}
+	})
+
+	By("Wait for test pods to be ready")
+	for _, obj := range objects {
+		pod, ok := obj.(*corev1.Pod)
+		if !ok {
+			continue
+		}
+
+		Eventually(func(g Gomega) {
+			g.Expect(c.Get(ctx, client.ObjectKeyFromObject(pod), pod)).To(Succeed())
+			g.Expect(pod.Status.Phase).To(Equal(corev1.PodRunning))
+		}).WithContext(ctx).WithTimeout(time.Minute).Should(Succeed(), "%T %q should get ready", obj, client.ObjectKeyFromObject(obj))
+	}
+}
+
+func execute(ctx context.Context, clientSet kubernetes.Interface, podName string, command ...string) *Buffer {
+	GinkgoHelper()
+	var stdOutBuffer, stdErrBuffer = NewBuffer(), NewBuffer()
+
+	Expect(clientSet.PodExecutor().ExecuteWithStreams(
+		ctx,
+		namespace,
+		podName,
+		containerName,
+		nil,
+		io.MultiWriter(stdOutBuffer, gexec.NewPrefixedWriter("[out] ", GinkgoWriter)),
+		io.MultiWriter(stdErrBuffer, gexec.NewPrefixedWriter("[err] ", GinkgoWriter)),
+		command...,
+	)).To(Succeed())
+
+	return stdOutBuffer
+}
+
+func getObjects(kubernetesVersion string) []client.Object {
+	objects := getRBACObjects()
+
+	image, err := imagevector.Containers().FindImage(imagevector.ContainerImageNameHyperkube, imagevectorutils.TargetVersion(kubernetesVersion))
+	Expect(err).NotTo(HaveOccurred())
+
+	podDirect := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      podNameDirect,
+			Namespace: namespace,
+			Labels:    maps.Clone(labels),
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{{
+				Name:  containerName,
+				Image: image.String(),
+				Env: []corev1.EnvVar{{
+					// disable color output of `kubectl cluster-info` to make it simpler to assert
+					// ref https://github.com/daviddengcn/go-colortext/blob/v1.0.0/ct_ansi.go#L12
+					Name:  "TERM",
+					Value: "dumb",
+				}},
+			}},
+			ServiceAccountName: name,
+		},
+	}
+	objects = append(objects, podDirect)
+
+	podAPIServerProxyIP := podDirect.DeepCopy()
+	podAPIServerProxyIP.Name = podNameAPIServerProxyIP
+	// disable gardener's injection of the KUBERNETES_SERVICE_HOST env var
+	podAPIServerProxyIP.Labels[resourcesv1alpha1.KubernetesServiceHostInject] = "disable"
+	objects = append(objects, podAPIServerProxyIP)
+
+	podAPIServerProxyHostname := podDirect.DeepCopy()
+	podAPIServerProxyHostname.Name = podNameAPIServerProxyHostname
+	// manually set the KUBERNETES_SERVICE_HOST env var, gardener does not overwrite it if present
+	container := podAPIServerProxyHostname.Spec.Containers[0]
+	container.Env = append(container.Env, corev1.EnvVar{
+		Name:  "KUBERNETES_SERVICE_HOST",
+		Value: "kubernetes.default.svc.cluster.local",
+	})
+	podAPIServerProxyHostname.Spec.Containers[0] = container
```

Comment:
Why not set it directly?

```suggestion
	podAPIServerProxyHostname.Spec.Containers[0].Env = append(podAPIServerProxyHostname.Spec.Containers[0].Env, corev1.EnvVar{
		Name:  "KUBERNETES_SERVICE_HOST",
		Value: "kubernetes.default.svc.cluster.local",
	})
```